frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3956,6 +4011,145 @@ bool isTypeDefaultInitializable(Context* context, const Type* t) {
   return isTypeDefaultInitializableQuery(context, t);
 }
 
+void getCopyOrAssignableInfo(Context* context, const Type* t,
+                                    bool* fromConst, bool* fromRef,
+                                    bool checkCopyable);
+
+// checkCopyable is true for copyable, false for assignable.
+// Return: First is from const, second is from ref.
+static const std::pair<bool, bool>& getCopyOrAssignableInfoQuery(
+    Context* context, const Type* t, bool checkCopyable) {
+  QUERY_BEGIN(getCopyOrAssignableInfoQuery, context, t, checkCopyable);
+
+  CHPL_ASSERT((t->isClassType() || t->isCompositeType()) &&
+              "should only be getting called for class or composite types");
+
+  // Inspect type for either kind of copyability/assignability.
+  bool fromConst = false;
+  bool fromRef = false;
+  if (auto classTy = t->toClassType()) {
```

Comment:
IMO it'd be better to leave this part out of the query, and only run the query for records. The rationale is that there's minimal work to do for the classes, and it's not worth storing the result for each different class type.